design/sticky-session-design.md

```diff
@@ -0,0 +1,343 @@
+# Sticky session support
+
+**Status**: _draft_
+
+This document proposes adding support for sticky session in contour.
+
+## Goals
+
+- Enabling support for sticky-sessions in Contour using
+    - Cookie
+    - Source IP
+    - Header
+- Adding support for sticky-sessions in IngressRoute CRD. 
+
+## Non-goals
+
+- Making options in IngressRoute very specific to Envoy.
+
+## Background
+
+Sticky-sessions allow requests to be routed to the same host based on certain parameters of the request.
+The parameters can be a cookie, the source IP, a custom header, etc.
+
+#### Sticky sessions in Envoy 
+Envoy supports sticky sessions by using a hash-based load balancing strategy (RingHash, Maglev). 
+Specifying a **hash policy** for such load-balancers will enable sticky sessions. Hash policy 
+is specified per route. **For sticky sessions, each route should only have one service
+associated to the route**. This is because if multiple clusters are associated per route (weighted clusters),
+requests will be forwarded to the cluster based on weights and stickiness will be 
+enforced in that cluster from that point. This defeats the purpose of using sticky sessions since a session will not
+be pinned to a pod. 
+
+## High-Level Design
+In contour, sticky sessions will be enabled per service. 
+If a service has sticky sessions enabled, it should be the **only service**
+under that route. Sticky sessions **do not** work with multiple services (weighted clusters).
+This means that sticky sessions cannot be used along with blue/green deployments.
+
+### Sticky-sessions configuration
+To enable sticky-sessions in a service, the `strategy` needs to be set to *StickySession*.
+By default, sticky-session will be based on **source-ip**. `stickySessionConfig` can 
+be provided in the service to specify the parameter for stickiness.
+
+#### Cookie-based stickiness
+This allows a request to be routed based on a cookie in the http request.
+The following parameters can be specified under the strategy options for cookie-based stickiness:
+- **name**: The name of the cookie to look for to determine stickiness.
+- **path**: The path for the cookie.
```

Comment:
What is this `path` parameter used for?

Reply:
It is used by envoy to set a path for the [cookie](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto.html?highlight=strip%20headers#envoy-api-msg-route-routeaction-hashpolicy-cookie). This is used when no cookie of the specified name is found in the request. In that case, envoy will generate a cookie of that name under the specified "path".

Reply:
Now that I think of it, if we support this in contour, then we should also include the TTL field. If a cookie is not found, a new cookie will be generated by envoy with the TTL and path set.

Reply:
Sure. It is an optional feature provided by Envoy to set the cookie in case it is missing. We can use [passive](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto.html?highlight=strip%20headers#route-routeaction-hashpolicy-cookie) cookie affinity to disable this behaviour.